### ISSUE TITLE
Galaxy API reference

### ISSUE THREAD

**Issue by Talv** [SC2Mapster/sc2mapster.github.io#4]
Like on the old wiki:
https://sc2mapster.gamepedia.com/Triggers/Functions

It should be auto-generated from SC2 data files. So we can keep it uptodate.

**Comment by Talv** [comment]
Self-assinging this one.
I already have scripts for building an index: https://github.com/Talv/plaxtony
So it's _mostly_ a matter of generating actual markdown files.

**Comment by abrahamYG** [comment]
I would suggest implwmenting this in the gamedata repo. I thibk it would be easy to generate the jenkings page. Do we want each function to have its own page?

**Comment by Talv** [comment]
I think having dedicated page might be beneficial - I thought about adding *Usage* section where it would list where certain function (or any other symbol) was used, in scope of default mods. (Also maybe *Related* section - where it would list other functions from the same category).
Other concern is whether we want to include non-natives. I mean stuff from libraries in default mods - campaign, allied commanders etc.

Also, I haven't yet worked with Jekyll and also no idea about current integration with Jenkins. If I add all these doc pages will it the the build times from the integration? If so, is there a way to exclude certain stuff?
I'd prefer to do the generation on my side, and push already baked html files.

I'll try to make some progress this week.

**Comment by abrahamYG** [comment]
> I think having dedicated page might be beneficial - I thought about adding Usage section where it would list where certain function (or any other symbol) was used, in scope of default mods. (Also maybe Related section - where it would list other functions from the same category).

I dig the idea of the usage section. Some of the functions are not terribly obvious abotu what they do until you see them in use.

>Other concern is whether we want to include non-natives. I mean stuff from libraries in default mods - campaign, allied commanders etc.

I would say no. While there are some nifty functions to learn from them, it might just lead to grow this into an undesired scale.

>Also, I haven't yet worked with Jekyll and also no idea about current integration with Jenkins. If I add all these doc pages will it the the build times from the integration? If so, is there a way to exclude certain stuff?
>I'd prefer to do the generation on my side, and push already baked html files.

Jekyll is a ruby gem you can install. Then basically you run it on your directory where you got the website docs, and it will host a web server that will automatically generate the baked files as you modify them. My build time for the website is around 11 seconds, but I have a slow cpu clock (1Ghz). There are config files to set up what gets rendered and what doesnt, plus you can set a template and override as you see fit.

**Comment by Talv** [comment]
Alright, I've figured Jekyll, but as a result I've come to conclusion that this might not be the best place for API docs generated from code.
- it would clutter the repo too much [1]. since it started as tutorials, let's leave it that way
- usual way of doing things in Jekyll wouldn't make this whole job easier, as I'd need to create separate markdown file for every page - preferably putting everything into YAML medata section, for customized rendering via templates.
- current theme is not suitable for this usecase, and I don't want to do an overhaul here

[1] In fact, repo is already way too big than it should, since you've pushed `/_site` directory, which should not be there. ;P And it at least doubled the size.

----

So I think it will be best, if I start building API docs in separate repo (in Mapster group ofc).

I've been looking around for static page generators designed to building API from code (via PluginAPI), but haven't quite found a favourite. It is quite likely I'll just built a custom one, even though I dislike reinventing the wheel :P
Though, considering that I can still use open source templates as a base, and there are drop-in modules for stuff like code colorizing, it should not be all that much work. And at least it will be flexible. What is important, since I wanna have a window in case I get around building docs for Data and UI as well. And perhaps some other customizations specific to SC2.

Anyways, this is what I've got so far:
[markdown example](https://gist.github.com/Talv/6cf249064995535c6d36e0f1855d1f65)
[json store](https://gist.github.com/Talv/7c1ea483e3cb2326368b87cede8b90ee)
*(Github truncated the files - as they're quite large, haha)*

**Comment by abrahamYG** [comment]
>it would clutter the repo too much

Well yeah, thats why I suggested using the [gamedata repo](https://github.com/SC2Mapster/SC2GameData). 

>[1] In fact, repo is already way too big than it should, since you've pushed /_site directory, which should not be there. ;P And it at least doubled the size.

Bite me haha.

What the issue is for making lots of markdown files.  I feel it would make sense to have an url based on the api fucntion tbh. We could categorize the api lists on galaxy/ layout/ and data/ (<--this one might need tons of sub categories tho).

Anyway showuld we mov ehtis disucssion to taht another repo?

**Comment by Talv** [comment]
I meant the clutter from all these auto generated pages living alongside the tutorials.

Anyway, I gave it a go nonetheless:
https://github.com/Talv/sc2mapster.github.io/commit/8ba6168a9f59c417a3497b6de4f91978a287adf3

And unfortunately it doesn't really work - Jekyll build process takes too long. Since I made these pages as collections, it grabs the Front Matter from markdown files at every change (in order to have it available as global variables from any page).

Not to mention ~20s build times when changing single thing in template layout tied to Galaxy... It's kinda understandable, but it doesn't make it easy to preview changes. I'd want it to only regenerate the page I'm requesting while in `serve` mode, but I don't think it can be done, due to how Jekyll is designed.

So yeah - I'll just make custom generator.

**Comment by abrahamYG** [comment]
I keep mentioning to move the api docs to a different repo, and I keep suggesting using this one: https://github.com/SC2Mapster/SC2GameData

I agree having it spearate from rgeular docs.

I ahvent taken a look at sphynx, any experience?

**Comment by Talv** [comment]
Ah, I thought you meant like having scripts responsible for generation, or intermediate YAML/JSON storage files in there. Not actual website/already build `.html` files. I'd rather have them in totally new repo.

I did take a rough look at Sphinx and I suppose it could work, but I'd likely be forced to write some plugins. And overall the hassle regarding figuring new framework/tech for relatively simple thing puts me back.

I kinda already have my set of tools I'm familiar with so I'll just go this way, it's not a rocket science to build static page generator anyway. Especially when using all the shit `npm` has to offer, haha. In fact I already started, maybe will have something ready today/tomorrow. I'll let you know then.
